### ISSUE TITLE
Default values in columns behaviour changed

### ISSUE THREAD

**Issue by brunosiqueira** [typeorm/typeorm#1892]
**Issue type:**

[ ] question
[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[x] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `websql`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**
I was using version **0.0.20** and I just migrated to version **0.1.19**. My User model has a language column with default value 'pt-BR'.

```
@Entity()
export class User {
  @PrimaryColumn('uuid')
  @IsUUID('4')
  id: string;

  @Column({ type: 'varchar', unique: true })
  username: string;

  @Column({type: 'varchar',
    length: 11,
    default: 'pt-br'
  })
  language: string;
}
```
I also had a automated test that was passing properly:

 ```
 it('should set null language to pt', (done) => {
    const user = new UserFactory().generate({ generatePassword: true });
    user.role = role
    user.sector = savedSector;
    delete user.language;
    userRepository
    .save(user)
    .then((newUser) => {
      expect(newUser.language).to.be.eq('pt-BR');
      done();
    }).catch((error) => done(error));
  });
```
After I upgraded to the latest version, my assertion started failing.
``` 
AssertionError: expected undefined to equal 'pt-BR'
```
I notice that the default generation changed, setting the value directly in the column. I think that it changed also the way the promise after the entity returns the populated object.

If I force to reload the object from the database it works properly:

```
  it('should set null language to pt', (done) => {
    const user = new UserFactory().generate({ generatePassword: true });
    user.role = role;
    user.sector = savedSector;
    delete user.language;
    userRepository
    .save(user)
    .then((newUser) => {
        return userRepository
          .findOneById(newUser.id)
          .then((newUser) => {
            expect(newUser.language).to.be.eq('pt-br');
            done();
      }).catch((error) => done(error));
    });
  });
```
So, there are two points of attention:
1 - the return from the save() method should load all the columns from the database after it is saved (or check the default values defined?)
2 - regarding compatibility, it is important to check how the new version of typeorm works with old generated databases. My previous one, already in production did not have the default value set in the database. Only in the model. And it also generates a null error if I run the tests with the old schema.

**Comment by pleerock** [comment]
Don't migrate to `0.1.19`. Migrate directly to `0.2.0` (npm i typeorm@next) - 0.2.0 is already much more stable and will be released very soon.

> I notice that the default generation changed, setting the value directly in the column.

yes if your model does not provide a value then database column's DEFAULT will be used.

> the return from the save() method should load all the columns from the database after it is saved (or check the default values defined?)

it loads only some columns like those which have DEFAULTs

>  regarding compatibility, it is important to check how the new version of typeorm works with old generated databases. My previous one, already in production did not have the default value set in the database. Only in the model. And it also generates a null error if I run the tests with the old schema.

upgrade to `next`, there are few bugs fixed with how `default` is updated in the database and provide more details on your issue.

**Comment by pleerock** [comment]
I think I answered all your questions and this issue can be closed. 0.2.0 is released btw.

**Comment by brunosiqueira** [comment]
I believe we should reopen this issue. I just tested my application with 0.2.5 version and I faced other issues with default value.

```
  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;
```

My model is being inserted without updatedAt and createdAt values. I realised that in this new version, those columns have the default value directly in the database ("now()") - but my database was generated from the previous version (without the default value). So it generates a null value on a required field error.

I also tried to automatically generate a migration with the differences, but it apparently does not identify new default values:

> No changes in database schema were found - cannot generate a migration. To create a new empty migration use "typeorm migration:create" command

Any thoughts on this? I think typeorm should still fallback gracefully to work with with old-version-generated-databases.
